footbot/research/utils/simulator.py

```diff
@@ -0,0 +1,406 @@
+
+    first_team_dicts = results_df.loc[
+        results_df["element_all"].isin(first_team), :
+    ].to_dict(orient="records")
+    bench_dicts = results_df.loc[results_df["element_all"].isin(bench), :].to_dict(
+        orient="records"
+    )
+
+    return first_team_dicts, bench_dicts
+
+
+def set_event_state(event, first_team, bench, bank, transfers_made, elements_df):
+    if event == 1:
+        existing_squad = []
+        total_budget = 1000
+        free_transfers_available = 1
+    else:
+        existing_squad = first_team + bench
+        team_value = calculate_team_value(first_team + bench, elements_df)
+        total_budget = team_value + bank
+        free_transfers_available = 2 if transfers_made == 0 else 1
+
+    return existing_squad, total_budget, free_transfers_available
+
+
+def make_transfers(
+    event,
+    events_to_look_ahead,
+    existing_squad,
+    total_budget,
+    first_team_factor,
+    bench_factor,
+    captain_factor,
+    vice_factor,
+    transfer_penalty,
+    transfer_limit,
+    predictions_df,
+    elements_df,
+):
+    players = get_team_selector_input(
+        predictions_df, elements_df, event, event + events_to_look_ahead
+    )
+
+    # when making a team from scratch we require 15 transfers
+    if event == 1:
+        transfer_limit = 15
+
+    first_team, bench, _, _, transfers = select_team(
+        players,
+        optimise_key="avg_predicted_total_points",
+        existing_squad=existing_squad,
+        total_budget=total_budget,
+        first_team_factor=first_team_factor,
+        bench_factor=bench_factor,
+        captain_factor=captain_factor,
+        vice_factor=vice_factor,
+        transfer_penalty=transfer_penalty,
+        transfer_limit=transfer_limit,
+    )
+
+    existing_squad = first_team + bench
+
+    team_value = calculate_team_value(first_team + bench, elements_df)
+
+    bank = total_budget - team_value
+
+    return existing_squad, bank, transfers
+
+
+def make_team_selection(
+    event,
+    existing_squad,
+    total_budget,
+    first_team_factor,
+    bench_factor,
+    captain_factor,
+    vice_factor,
+    predictions_df,
+    elements_df,
+):
+    players = get_team_selector_input(predictions_df, elements_df, event, event)
+
+    first_team, bench, captain, vice, _ = select_team(
+        players,
+        optimise_key="avg_predicted_total_points",
+        existing_squad=existing_squad,
+        total_budget=total_budget,
+        first_team_factor=first_team_factor,
+        bench_factor=bench_factor,
+        captain_factor=captain_factor,
+        vice_factor=vice_factor,
+        transfer_penalty=0,
+        transfer_limit=0,
+    )
+
+    return first_team, bench, captain, vice
+
+
+def simulate_event(
+    season,
+    event,
+    purchase_price_dict,
+    predictions_df,
+    first_team,
+    bench,
+    bank,
+    transfers_made,
+    events_to_look_ahead,
+    first_team_factor,
+    bench_factor,
+    captain_factor,
```

Comment:
this function probably has a few too many arguments.. since other functions share a similar signature it might be a bit neater to have some sort of 'config' object that holds the various factors, input settings

Reply:
if you don't want to do this, an easy compromise is to force all arguments to be named:

```
def my_function(
  *, <- consumes all unnamed arguments
  first_kwarg,
  second_kwarg,
)
```

so the caller always has to name the args. otherwise its almost impossible to see bugs where you specify args in wrong order

Reply:
I've gone for the latter, just for simplicity. But if it becomes a pain point we can refactor.